cli/src/file_access.rs

```diff
@@ -0,0 +1,90 @@
+use crate::Args;
+use modmark_core::AccessPolicy;
+use std::collections::HashMap;
+use std::io::{stdin, stdout, Write};
+use std::path::Path;
+
+#[derive(Default)]
+struct ModulePermissions {
+    permissions: HashMap<String, bool>,
+}
+
+pub struct CliAccessManager {
+    root: String,
+    deny_read: bool,
+    deny_write: bool,
+    deny_create: bool,
+    allow_every_module: bool,
+    modules: HashMap<String, ModulePermissions>,
+}
+
+impl AccessPolicy for CliAccessManager {
+    fn root(&self) -> String {
+        self.root.clone()
+    }
+
+    fn allowed_to_read(&self) -> bool {
+        !self.deny_read
+    }
+
+    fn allowed_to_write(&self) -> bool {
+        !self.deny_write
+    }
+
+    fn allowed_to_create(&self) -> bool {
+        !self.deny_create
+    }
+
+    fn allowed_access(&mut self, path: &Path, module_name: &String) -> bool {
+        if self.allow_every_module {
+            return true;
+        }
+
+        let module = self.modules.entry(module_name.clone()).or_default();
+        let path_str = path.to_str().expect("Could not convert path to &str");
+
+        if let Some(access) = module.permissions.get(path_str) {
+            *access
+        } else {
+            // If we have not asked the user before, prompt them
+            let prompt =
+                format!("\nModule [{module_name}] requests access to \"{path_str}\" (y/n): ");
+            let result = prompt_user(&prompt);
+            module.permissions.insert(path_str.to_string(), result);
+            result
+        }
+    }
+}
+
+// TODO LATER: make async to work better with the rest of the CLI
+fn prompt_user(prompt: &str) -> bool {
+    println!("{prompt}");
+
+    let allowed_input = ["y", "n", "Y", "N"];
+
+    loop {
+        let mut buffer = String::new();
+        stdin().read_line(&mut buffer).expect("");
+
+        let input = buffer.trim();
+        if allowed_input.contains(&input) {
+            return input == "y" || input == "Y";
+        }
+
+        print!("Unexpected input. Please enter (y/n): ");
+        stdout().flush().expect("Could not flush output");
+    }
+}
+
+impl CliAccessManager {
+    pub(crate) fn new(args: &Args) -> Self {
+        Self {
+            root: args.assets.clone().unwrap_or(String::from("assets")),
```

Comment:
If no `--assets` flag was specified we assume that the user wants to open a "assets" directory. I think that we instead should skip the preopen step and not open a directory at all. This way, we would be more secure by default and with the current system it is impossible to compile documents with exposing a directory:

If we compile without having a assets sub directory we get this error:
![image](https://user-images.githubusercontent.com/11508459/228758650-319ff9d6-9b97-45b0-995c-ee87012b4efd.png)

and we add  `--deny-read --deny-create --deny-write` all of these files even if we have a valid directory we get the following error:
 
![image](https://user-images.githubusercontent.com/11508459/228759033-187d1123-1054-4193-bc92-b76566a600f6.png)

Reply:
Fixed, I believe! I was unable to find a neat way to introduce conditional execution in the call chain for `WasiStateBuilder`. I tried containing the if statements to the preopen-closure, but the closure is practically required to give a valid path to the `PreopenDirBuilder`.

Reply:
This commit also deals with second screenshot. If read, write and create all are false, it will not try to preopen a directory.